dodos/forecast.py

```diff
@@ -69,15 +76,42 @@ def task_forecast_predict():
     Forecast: produce predictions for the given time range.
     """
 
-    def forecast_action(time_start, time_end):
+    def forecast_action(pred_start, pred_end, pred_horizon, pred_interval, pred_seqlen):
+        # Read the query log timestamps from the preprocessor's output.
+        # TODO(WAN): This file is read repeatedly!
+        with open(PREPROCESSOR_TIMESTAMP) as ts_file:
+            lines = ts_file.readlines()
+            assert len(lines) == 2, "Timestamp file should have two lines with a timestamp each."
+            log_start = pd.Timestamp(lines[0]).floor(pred_interval)
+            log_end = pd.Timestamp(lines[1]).floor(pred_interval)
+
+        # Infer the prediction window from the query log and default parameters.
+        if pred_start is None:
+            pred_start = log_end + pred_interval
+        if pred_end is None:
+            pred_end = log_end + pred_horizon
```

Comment:
Actually, did you intend this to be `log_end + pred_interval + pred_horizon`?

Reply:
@mkpjnx Will merge pending resolution of this comment.

Reply:
Yes. We want to predict intervals up to `log_end + pred_horizon`, which means that that the end of the window for prediction is (`log_end+pred_horizon+pred_interval`)